File: 5-Problem-A-Day/92. Reverse Linked List II.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; } }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = dummy;
        int count = 0;
        ListNode prev = null, follow = null, before = null, last = null;
        while (curr != null) {
            if (count >= m && count <= n) {
                ListNode tmp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = tmp;
            } else {
                if (count == m - 1) {
                    before = curr;
                    last = curr.next;
                } else if (count == n + 1)
                    follow = curr;
                curr = curr.next;
            }
            count += 1;
        }
        before.next = prev;
        last.next = follow;
        return dummy.next;
    }
}

File: 2021/92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* dummy = new ListNode();
        dummy->next = head;
        ListNode* leftPrev = dummy;
        for (int i = 1; i < left; i++) {
            leftPrev = leftPrev->next;            
        }
        ListNode *curr = leftPrev->next, *prev = NULL, *rightNext = NULL, *leftLast = curr;                
        for (int i = left; i <= right; i++) {
            rightNext = curr->next;
            curr->next = prev;            
            prev = curr;
            curr = rightNext;
        }
        leftLast->next = rightNext;
        leftPrev->next = prev;
        head = dummy->next;
        delete dummy;
        return head;        
    }
};